contracts/StablecoinConverter.sol

```diff
@@ -0,0 +1,119 @@
+pragma solidity ^0.5.0;
+
+import "./EpochTokenLocker.sol";
+import "./libraries/IdToAddressBiMap.sol";
+import "solidity-bytes-utils/contracts/BytesLib.sol";
+import "./libraries/IdToAddressBiMap.sol";
+
+
+contract StablecoinConverter is EpochTokenLocker {
+    using SafeMath for uint;
+    using BytesLib for bytes32;
+
+    event OrderPlacement(
+        address owner,
+        uint16 buyToken,
+        uint16 sellToken,
+        bool sellOrderFlag,
+        uint32 validFrom,
+        uint32 validUntil,
+        uint256 buyAmount,
+        uint256 sellAmount
+    );
+
+    event OrderCancelation(
+        address owner,
+        uint id
+    );
+
+    //outstanding volume of an order is encoded as buyAmount or sellAmount depending on sellOrderFlag
+    struct Order {
+        uint16 buyToken;
+        uint16 sellToken;
+        uint32 validFrom;  // order is valid from validFrom inclusively
+        uint32 validUntil;  // order is valid till validUntil inclusively
+        bool sellOrderFlag;
+        uint128 buyAmount;
+        uint128 sellAmount;
+    }
+
+    // User-> Order
+    mapping(address => Order[]) public orders;
+
+    IdToAddressBiMap.Data private registeredTokens;
+    uint public MAX_TOKENS;
+    uint16 public numTokens = 0;
+
+    constructor(uint maxTokens) public {
+        MAX_TOKENS = maxTokens;
+    }
+
+    function addToken(address _tokenAddress) public {
+        require(numTokens < MAX_TOKENS, "Max tokens reached");
+        require(
+            IdToAddressBiMap.insert(registeredTokens, numTokens, _tokenAddress),
+            "Token already registered"
+        );
+        numTokens++;
+    }
+
+    function placeOrder(
+        uint16 buyToken,
+        uint16 sellToken,
+        bool sellOrderFlag,
+        uint32 validUntil,
+        uint128 buyAmount,
+        uint128 sellAmount
+    ) public returns (uint) {
+        orders[msg.sender].push(Order({
+            buyToken: buyToken,
+            sellToken: sellToken,
+            validFrom: getCurrentStateIndex() + 1,
+            validUntil: validUntil,
+            sellOrderFlag: sellOrderFlag,
+            buyAmount: buyAmount,
+            sellAmount: sellAmount
+        }));
+        emit OrderPlacement(
+            msg.sender,
+            buyToken,
+            sellToken,
+            sellOrderFlag,
+            getCurrentStateIndex() + 1,
+            validUntil,
+            buyAmount,
+            sellAmount
+        );
+        return orders[msg.sender].length - 1;
+    }
+
+    function cancelOrder(
+        uint id
+    ) public {
+        orders[msg.sender][id].validUntil = getCurrentStateIndex();
+        emit OrderCancelation(msg.sender, id);
+    }
+
+    function freeStorageOfOrder(
+        uint id
+    ) public {
+        require(orders[msg.sender][id].validUntil < getCurrentStateIndex(), "Order is still valid");
+        orders[msg.sender][id] = Order({
```

Comment:
Why not using `delete`?

Reply:
Because delete would shift indexes for all other orders. And then solution submission in-flight might no longer be correct

Reply:
Are you sure? 

I made this small test and it looks like delete only add 0s to the position you delete:

```
pragma solidity ^0.5.0;

contract TestCancelOrder {
    struct Order{
        string name;
    }
    
    Order[] private orders;
    
    function submitOrder() public {
        orders.push(
            Order("buy A for B :)")
        );
    }
    
    function orderById(uint id) public view returns(string memory res) {
        return orders[id].name;
    }
    
    function cancelOrder(uint id) public {
        return delete orders[id];
    }
    
    function numOrders() public view returns(uint count) {
        return orders.length;
    }
}
```

- Test to push `submitOrder` some times
- Delete one item in the middle
- Check the count + try to get the position of the deleted item
- Try to delete the last item in the array
- Check the count + try to get the position of the deleted item

Reply:
Yeah, you are right. Good catch:
https://ethereum.stackexchange.com/questions/1527/how-to-delete-an-element-at-a-certain-index-in-an-array/39302

Reply:
Very cool and useful discussion here! Great find @anxolin!